README.md

```diff
@@ -97,60 +97,65 @@ let encodedTrifleSignedDataProto = try trifleSignedData.serialize()
 ```kotlin
 // Check if a key already exists.
 // If no key exists, generate a public key pair
-var keyHandle =  TrifleApi.generateKeyHandle(alias: abc)
+var keyHandle: KeyHandle =  TrifleApi.generateKeyHandle("abc")
                 
-// Storing keys. Keys are codable.
-val encoder = Gson()
-let jsonKeyHandle = try encoder.toJson(keyHandle)
+// Storing keys. Key handles are serializable.
+val encoded: ByteArray = keyHandle.serialize()
 
 // Load the key from storage when we need to use it
+val decoded: KeyHandle = KeyHandle.deserialize(encoded)
 
 // Check the validity of loaded key
-// TBD trifleApi.isValid(keyHandle: keyHandle)
+val valid: Boolean = trifleApi.isValid(keyHandle)
 
 // Destroy key that is no longer in use or is invalid
-// TBD trifleApi.delete(keyHandle: keyHandle)
+trifleApi.delete(keyHandle)
         
 // Check if loaded key already has a cert. If yes, skip to checking for cert validity
 // Else if key does not have a cert OR if a new cert must be generated (eg because of existing
 // cert is already expired, or app needs to re-attest, app is re-installed, app is restored
 // from backup, ... etc)
 
-// Create cert request
-val certReq = TrifleApi.generateMobileCertificateRequest(entity = entity, keyHandle = keyHandle)
+// Create cert request with an entity identity that is associated with the public key
+val certReq: CertificateRequest = TrifleApi.generateMobileCertificateRequest(entity, keyHandle)
 
-// Serialize to proto to be sent over wire
+// Serialize as an opaque ByteArray
+val encoded: ByteArray = certReq.serialize()
 
-// Send certificate request to Certificate Authority endpoint. Response will be [Data]
-// Iterate over each Data to convert to TrifleCertificate
+// Send certificate request to Certificate Authority endpoint. Response will be List<ByteArray>
+val response: List<ByteArray>
 
-// certs is an array of certificates where [0] will be device certificate
+// Iterate over each ByteArray to convert to Certificate
+val certs: List<Certificate> = response.map { Certificate.deserialize(it) }
+
+// certs is a list of certificates where [0] will be device certificate
 // and the rest of the elements will be intermediate chain.
 
 // Check if app has the root cert of Certificate Authority (CA). 
 
 // Validate cert matches the certificate request (so generated key) 
 // and the root (so it has been generated by the right CA).
-// TBD var isValid = TrifleApi.verify(csr, certs, anchor)
+val result: Result<Unit> = certs[0].verify(certReq, certs, root)
```

Comment:
Can we document a list of the exceptions that verify can throw that apps need to catch? specifically about bad anchor,  bad sigature, expiration etc.

Reply:
done